### PR TITLE
time: Replace struct timeval with scalar value

### PR DESCRIPTION
**Pull request by jlucovsky**
Issue: 5718

This commit switches the majority of time handling to a new type -- SCTime_t -- which is a 64 bit container for time:
- 44 bits -- seconds
- 20 bits -- useconds

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [5718](https://redmine.openinfosecfoundation.org/issues/5718)

Describe changes:
- Replace packet, thread, flow and other places using `struct timeval` with `SCTime_t`
- SCTime_t values are never passed by reference; always by value
- 64-bit SCTime_t value can represent over 500,000 years.

#suricata-verify-pr:
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .app_layer.error.tls.parser | 931 | 1145 | 122.99% |
| | SURI_TLPR1_stats_chk | | |
| .decoder.event.ipv4.frag_overlap | 805 | 5 | 0.62% |
| .flow.memuse | 565135168 | 634841584 | 112.33% |
| .app_layer.flow.http | 3118951 | 3340799 | 107.11% |
| .app_layer.flow.tls | 510462 | 623355 | 122.12% |
| .app_layer.flow.ssh | 8937 | 11032 | 123.44% |
| .app_layer.flow.telnet | 1624 | 2334 | 143.72% |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 139587 | 0.02 | 

Pipeline 11604

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128237 | 104171 | 81.23% |
| .app_layer.error.tls.parser | 931 | 1145 | 122.99% |
| | SURI_TLPR1_stats_chk | | |
| .decoder.event.ipv4.frag_overlap | 805 | 5 | 0.62% |
| .app_layer.flow.http | 3118951 | 3340799 | 107.11% |
| .app_layer.flow.tls | 510462 | 623355 | 122.12% |
| .app_layer.flow.ssh | 8937 | 11032 | 123.44% |
| .app_layer.flow.telnet | 1624 | 2334 | 143.72% |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 11502 | 0.00 | 

Pipeline 11605

File: src/flow-util.h [hunk 1]
```diff
@@ -36,96 +36,96 @@
...
-        (f)->de_ctx_version = 0; \
-        (f)->thread_id[0] = 0; \
-        (f)->thread_id[1] = 0; \
-        (f)->alparser = NULL; \
-        (f)->alstate = NULL; \
-        (f)->sgh_toserver = NULL; \
-        (f)->sgh_toclient = NULL; \
-        (f)->flowvar = NULL; \
-        RESET_COUNTERS((f)); \
+#define FLOW_INITIALIZE(f)                                                                         \
```
Comment (victorjulien) [inline src/flow-util.h:39]: would be nice to do this in a separate commit before the main one

File: src/tm-threads.c [hunk 2]
```diff
@@ -2231,17 +2230,21 @@ void TmThreadsGetMinimalTimestamp(struct timeval *ts)
...
+            .tv_usec = SCTIME_USECS(t->pktts) };
+        if (!(timercmp(&pkttv, &nullts, ==))) {
             /* ignore sleeping threads */
             if (t->sys_sec_stamp + 1 < (uint32_t)systs.tv_sec)
                 continue;
 
             if (!set) {
-                local = t->pktts;
-                set = 1;
+                local.tv_sec = SCTIME_SECS(t->pktts);
```
Comment (victorjulien) [inline src/tm-threads.c:2241]: use a macro for doing this in a single step?

File: src/util-debug.c [hunk 3]
```diff
@@ -667,8 +666,8 @@ SCError SCLogMessage(const SCLogLevel log_level, const char *file, const unsigne
     }
 
     /* get ts here so we log the same ts to each output */
-    struct timeval tval;
-    gettimeofday(&tval, NULL);
+    SCTime_t ts;
+    TimeGet(&ts);
```
Comment (victorjulien) [inline src/util-debug.c:670]: wonder if we should change this to `SCTime_t ts = TimeGet();`

File: src/app-layer-htp-range.c [hunk 4]
```diff
@@ -127,10 +127,10 @@ static void ContainerUrlRangeFree(void *s)
     }
 }
 
-static inline bool ContainerValueRangeTimeout(HttpRangeContainerFile *cu, struct timeval *ts)
+static inline bool ContainerValueRangeTimeout(HttpRangeContainerFile *cu, const SCTime_t *ts)
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/app-layer-htp-range.c:130]: Yes, I plan on doing that once CI is clean.

File: src/flow.c [hunk 5]
```diff
@@ -418,19 +418,19 @@ void FlowHandlePacketUpdate(Flow *f, Packet *p, ThreadVars *tv, DecodeThreadVars
     if (state != FLOW_STATE_CAPTURE_BYPASSED) {
 #endif
         /* update the last seen timestamp of this flow */
-        if (timercmp(&p->ts, &f->lastts, >)) {
-            COPY_TIMESTAMP(&p->ts, &f->lastts);
-            const uint32_t timeout_at = (uint32_t)f->lastts.tv_sec + f->timeout_policy;
+        if (SCTIME_CMP(p->ts, f->lastts, >)) {
```
[the remark this answers is not in the document]
Comment (jlucovsky) [reply src/flow.c:421]: Ok ... I'll add those macros.

File: src/util-time.h [hunk 6]
```diff
@@ -24,13 +24,41 @@
 #ifndef __UTIL_TIME_H__
 #define __UTIL_TIME_H__
 
+typedef uint64_t SCTime_t;
```
Comment (victorjulien) [inline src/util-time.h:27]: if we'd make this something like
```c
typedef struct {
  uint64_t secs:44;
  uint64_t usecs:20;
} SCTime_t;
```
We'd get masking for free I think (not at runtime, but code / review style overhead)

File: src/flow-util.h (hunk 1 above)
Comment (jlucovsky) [reply src/flow-util.h:39]: Which change?

The initialization of `lastts` changed; the whitespace is from `clang-format`.

File: src/util-time.h (hunk 6 above)
Comment (jlucovsky) [reply src/util-time.h]: Will do in a separate commit.

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128237 | 104197 | 81.25% |
| .app_layer.error.tls.parser | 931 | 1145 | 122.99% |
| | SURI_TLPR1_stats_chk | | |
| .app_layer.flow.http | 3118951 | 3340798 | 107.11% |
| .app_layer.flow.tls | 510462 | 623355 | 122.12% |
| .app_layer.flow.ssh | 8937 | 11032 | 123.44% |
| .app_layer.flow.telnet | 1624 | 2334 | 143.72% |
| .app_layer.error.ftp-data.gap | 0 | 1 | - |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 2441 | 0.00 | 

Pipeline 11615

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .tcp.rst | 128237 | 104330 | 81.36% |
| .app_layer.error.tls.parser | 931 | 1145 | 122.99% |
| | SURI_TLPR1_stats_chk | | |
| .flow.memuse | 565135168 | 873939712 | 154.64% |
| .app_layer.flow.http | 3118951 | 3340796 | 107.11% |
| .app_layer.flow.tls | 510462 | 623355 | 122.12% |
| .app_layer.flow.ssh | 8937 | 11032 | 123.44% |
| .app_layer.flow.telnet | 1624 | 2334 | 143.72% |
| .app_layer.error.ftp-data.gap | 0 | 1 | - |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 88236 | 0.01 | 

Pipeline 11632

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW1_stats_chk | | |
| .app_layer.error.tls.parser | 931 | 1145 | 122.99% |
| | SURI_TLPR1_stats_chk | | |
| .app_layer.flow.http | 3118951 | 3340798 | 107.11% |
| .app_layer.flow.tls | 510462 | 623355 | 122.12% |
| .app_layer.flow.ssh | 8937 | 11032 | 123.44% |
| .app_layer.flow.telnet | 1624 | 2334 | 143.72% |
| .app_layer.error.ftp-data.gap | 0 | 1 | - |
| | TREX_GENERIC_stats_chk | | |
| .capture.kernel_drops | 0 | 1673 | 0.00 | 

Pipeline 11634

File: src/flow-util.h (hunk 1 above)
Comment (victorjulien) [reply src/flow-util.h:39]: yeah the clang reformat as a separate commit before the lastts change

**Comment by jlucovsky** [comment]
Continued in #8396

File: src/flow-util.h (hunk 1 above)
Comment (jlucovsky) [reply src/flow-util.h:39]: Addressed in #8396
